File: frontend/src/types/lab.ts
export interface LabTask {
  id: string;
  description: string;
  language?: string;
}

export interface ParsedLab {
  title?: string;
  rawText: string;
  tasks: LabTask[];
}

export interface GeneratedTaskSolution {
  id: string;
  description: string;
  language: string;
  filename: string;
  code: string;
  explanation: string;
}

export interface GeneratedLab {
  labTitle: string;
  objectives: string[];
  tasks: GeneratedTaskSolution[];
  conclusion: string;
}

export type ExecutionStatus = "success" | "failed" | "timeout" | "unsupported";

export interface ExecutionResult {
  status: ExecutionStatus;
  stdout: string;
  stderr: string;
  exitCode?: number;
}

export type University = "air" | "bahria" | "nust";

File: backend/app/schemas/lab.py
"""Ported from src/types/lab.ts.

None of these are wired into the pipeline yet — same as in the original
repo, they exist so later phases (parser, AI adapter, executor, docx
generator) have an agreed shape to target.
"""

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class LabTask(BaseModel):
    id: str
    description: str
    language: Optional[str] = None


class ParsedLab(BaseModel):
    title: Optional[str] = None
    raw_text: str
    tasks: list[LabTask]


class GeneratedTaskSolution(BaseModel):
    id: str
    description: str
    language: str
    filename: str
    code: str
    explanation: str


class GeneratedLab(BaseModel):
    lab_title: str
    objectives: list[str]
    tasks: list[GeneratedTaskSolution]
    conclusion: str


class ExecutionStatus(str, Enum):
    SUCCESS = "success"
    FAILED = "failed"
    TIMEOUT = "timeout"
    UNSUPPORTED = "unsupported"


class ExecutionResult(BaseModel):
    status: ExecutionStatus
    stdout: str
    stderr: str
    exit_code: Optional[int] = None
